README
NAME
    Sub::Prototype::Util - Prototype-related utility routines.

VERSION
    Version 0.10

SYNOPSIS
        use Sub::Prototype::Util qw<flatten wrap recall>;

        my @a = qw<a b c>;
        my @args = ( \@a, 1, { d => 2 }, undef, 3 );

        my @flat = flatten '\@$;$', @args; # ('a', 'b', 'c', 1, { d => 2 })
        recall 'CORE::push', @args; # @a contains 'a', 'b', 'c', 1, { d => 2 }, undef, 3
        my $splice = wrap 'CORE::splice';
        my @b = $splice->(\@a, 4, 2); # @a is now ('a', 'b', 'c', 1, 3) and @b is ({ d => 2 }, undef)

DESCRIPTION
    Prototypes are evil, but sometimes you just have to bear with them,
    especially when messing with core functions. This module provides
    several utilities aimed at facilitating "overloading" of prototyped
    functions.

    They all handle 5.10's "_" prototype.

FUNCTIONS
  "flatten $proto, @args"
    Flattens the array @args according to the prototype $proto. When @args
    is what @_ is after calling a subroutine with prototype $proto,
    "flatten" returns the list of what @_ would have been if there were no
    prototype. It croaks if the arguments can't possibly match the required
    prototype, e.g. when a reference type is wrong or when not enough
    elements were provided.

  "wrap $name, %opts"
    Generates a wrapper that calls the function $name with a prototyped
    argument list. That is, the wrapper's arguments should be what @_ is
    when you define a subroutine with the same prototype as $name.

        my $a = [ 0 .. 2 ];
        my $push = wrap 'CORE::push';
        $push->($a, 3, 4); # returns 3 + 2 = 5 and $a now contains 0 .. 4

    You can force the use of a specific prototype. In this case, $name must
    be a hash reference that holds exactly one key / value pair, the key
    being the function name and the value the prototpye that should be used
    to call it.

        my $push = wrap { 'CORE::push' => '\@$' }; # only pushes 1 arg

    Others arguments are seen as key / value pairs that are meant to tune
    the code generated by "wrap". Valid keys are :

    "ref => $func"
        Specifies the function used in the generated code to test the
        reference type of scalars. Defaults to 'ref'. You may also want to
        use "reftype" in Scalar::Util.

    "wrong_ref => $code"
        The code executed when a reference of incorrect type is encountered.
        The result of this snippet is also the result of the generated code,
        hence it defaults to 'undef'. It's a good place to "croak" or "die"
        too.

    "sub => $bool"
        Encloses the code into a "sub { }" block. Default is true.

    "compile => $bool"
        Makes "wrap" compile the code generated and return the resulting
        code reference. Be careful that in this case "ref" must be a fully
        qualified function name. Defaults to true, but turned off when "sub"
        is false.

    For example, this allows you to recall into "CORE::grep" and "CORE::map"
    by using the "\&@" prototype :

        my $grep = wrap { 'CORE::grep' => '\&@' };
        sub mygrep (&@) { $grep->(@_) } # the prototypes are intentionally different

  "recall $name, @args"
    Calls the function $name with the prototyped argument list @args. That
    is, @args should be what @_ is when you call a subroutine with $name as
    prototype. You can still force the prototype by passing "{ $name =>
    $proto }" as the first argument.

        my $a = [ ];
        recall { 'CORE::push' => '\@$' }, $a, 1, 2, 3; # $a just contains 1

    It's implemented in terms of "wrap", and hence calls "eval" at each run.
    If you plan to recall several times, consider using "wrap" instead.

EXPORT
    The functions "flatten", "wrap" and "recall" are only exported on
    request, either by providing their name or by the ':funcs' and ':all'
    tags.

DEPENDENCIES
    Carp, Exporter (core modules since perl 5), Scalar::Util (since 5.7.3).

AUTHOR
    Vincent Pit, "<perl at profvince.com>", <http://www.profvince.com>.

    You can contact me by mail or on "irc.perl.org" (vincent).

BUGS
    Please report any bugs or feature requests to "bug-sub-prototype-util at
    rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Sub-Prototype-Util>. I
    will be notified, and then you'll automatically be notified of progress
    on your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc Sub::Prototype::Util

    Tests code coverage report is available at
    <http://www.profvince.com/perl/cover/Sub-Prototype-Util>.

COPYRIGHT & LICENSE
    Copyright 2008,2009,2010,2011 Vincent Pit, all rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Sub-Prototype-Util

0.10    2011-08-25 11:05 UTC
        + Chg : recall() now uses "goto" to call the wrapper code (except for
                perl 5.8.0).
        + Chg : The internals sustained a large cleanup. This will hopefully
                make it easier to understand and maintain.
        + Fix : The functions of this module no longer clobber $@.
        + Fix : Fix test failures on perl 5.8.0, and on perl 5.13.7 and above.
        + Fix : Work around Kwalitee test misfailures.

0.09    2009-06-04 22:45 UTC
        + Chg : recall() was reimplemented in terms of wrap().
        + Chg : "compile => 1" is now the default for wrap.
        + Fix : Croak messages are now relative to the caller.
        + Fix : flatten() croaks when not enough arguments were given to match
                the prototype.
        + Upd : Resources in META.yml.
        + Upd : META.yml spec updated to 1.4.

0.08    2008-05-02 15:25 UTC
        + Fix : "{ $func => undef }" (resp. "=> ''") should enforce no prototype
                (resp. enforce an empty prototype).

0.07    2008-04-21 09:00 UTC
        + Add : Forward eval() errors when compiling in wrap().
        + Add : Talk about wrap() in the synopsis and samples/try.pl.
        + Fix : t/12-wrap.t failures with 5.6.x.

0.06    2008-04-20 16:20 UTC
        + Add : The wrap() function.

0.05    2008-04-15 09:45 UTC
        + Add : You can now specify which prototype to use with recall(). This
                is needed to be able to recall into CORE::{grep,map}.

0.04    2008-04-09 13:25 UTC
        + Fix : recall() should pass by reference as much as possible, or we
                won't be able to apply it to functions that modify their
                arguments (e.g. open()).
        + Fix : our really doesn't exist in 5.005.

0.03    2008-04-06 22:20 UTC
        + Fix : our doesn't exist in 5.005 (sigh).
        + Fix : '_' prototype should use the current argument when it's
                available and only resort to $_ when it's not.

0.02    2008-04-06 16:40 UTC
        + Fix : Missing LICENSE in Makefile.PL.
        + Fix : Typos in POD.

0.01    2008-04-06 14:00 UTC
        First version, released on an unsuspecting world.



Makefile.PL
use 5.006;

use strict;
use warnings;
use ExtUtils::MakeMaker;

my $dist = 'Sub-Prototype-Util';

(my $name = $dist) =~ s{-}{::}g;

(my $file = $dist) =~ s{-}{/}g;
$file = "lib/$file.pm";

my %PREREQ_PM = (
 'Carp'         => 0,
 'Exporter'     => 0,
 'Scalar::Util' => 0,
);

my %META = (
 configure_requires => {
  'ExtUtils::MakeMaker' => 0,
 },
 build_requires => {
  'ExtUtils::MakeMaker' => 0,
  'Scalar::Util'        => 0,
  'Test::More'          => 0,
  %PREREQ_PM,
 },
 dynamic_config => 0,
 resources => {
  bugtracker => "http://rt.cpan.org/NoAuth/ReportBug.html?Queue=$dist",
  homepage   => "http://search.cpan.org/dist/$dist/",
  license    => 'http://dev.perl.org/licenses/',
  repository => "http://git.profvince.com/?p=perl%2Fmodules%2F$dist.git",
 },
);

WriteMakefile(
 NAME             => $name,
 AUTHOR           => 'Vincent Pit <perl@profvince.com>',
 LICENSE          => 'perl',
 VERSION_FROM     => $file,
 ABSTRACT_FROM    => $file,
 PL_FILES         => {},
 PREREQ_PM        => \%PREREQ_PM,
 MIN_PERL_VERSION => 5.006,
 META_MERGE       => \%META,
 dist             => {
  PREOP    => "pod2text -u $file > \$(DISTVNAME)/README",
  COMPRESS => 'gzip -9f', SUFFIX => 'gz'
 },
 clean            => {
  FILES => "$dist-* *.gcov *.gcda *.gcno cover_db Debian_CPANTS.txt"
 }
);


samples/try.pl
#!/usr/bin/env perl

use strict;
use warnings;

use Data::Dumper;

use lib qw{blib/lib};

use Sub::Prototype::Util qw/flatten recall wrap/;

my @a = qw/a b c/;
print "At the beginning, \@a contains :\n", Dumper(\@a);

my @args = ( \@a, 1, { d => 2 }, undef, 3 );
print "Our arguments are :\n", Dumper(\@args);

my $proto = '\@$;$';
my @flat = flatten $proto, @args; # ('a', 'b', 'c', 1, { d => 2 })
print "When flatten with prototype $proto, this gives :\n", Dumper(\@flat);

recall 'CORE::push', @args; # @a contains 'a', 'b', 'c', 1, { d => 2 }, undef, 3
print "After recalling CORE::push with \@args, \@a contains :\n", Dumper(\@a);

my $splice = wrap 'CORE::splice';
my @b = $splice->(\@a, 4, 2);
print "After calling wrapped splice with \@a, it contains :\n", Dumper(\@a);
print "What was returned :\n", Dumper(\@b);


lib/Sub/Prototype/Util.pm
package Sub::Prototype::Util;

use 5.006;

use strict;
use warnings;

use Carp         qw<croak>;
use Scalar::Util qw<reftype>;

=head1 NAME

Sub::Prototype::Util - Prototype-related utility routines.

=head1 VERSION

Version 0.10

=cut

use vars qw<$VERSION>;

$VERSION = '0.10';

=head1 SYNOPSIS

    use Sub::Prototype::Util qw<flatten wrap recall>;

    my @a = qw<a b c>;
    my @args = ( \@a, 1, { d => 2 }, undef, 3 );

    my @flat = flatten '\@$;$', @args; # ('a', 'b', 'c', 1, { d => 2 })
    recall 'CORE::push', @args; # @a contains 'a', 'b', 'c', 1, { d => 2 }, undef, 3
    my $splice = wrap 'CORE::splice';
    my @b = $splice->(\@a, 4, 2); # @a is now ('a', 'b', 'c', 1, 3) and @b is ({ d => 2 }, undef)

=head1 DESCRIPTION

Prototypes are evil, but sometimes you just have to bear with them, especially when messing with core functions.
This module provides several utilities aimed at facilitating "overloading" of prototyped functions.

They all handle C<5.10>'s C<_> prototype.

=head1 FUNCTIONS

=cut

my %sigils   = qw<SCALAR $ ARRAY @ HASH % GLOB * CODE &>;
my %reftypes = reverse %sigils;

sub _check_ref {
 my ($arg, $sigil) = @_;

 my $reftype;
 if (not defined $arg or not defined($reftype = reftype $arg)) {
  # not defined or plain scalar
  my $that = (defined $arg) ? 'a plain scalar' : 'undef';
  croak "Got $that where a reference was expected";
 }

 croak "Unexpected $reftype reference" unless exists $sigils{$reftype}
                                          and $sigil =~ /\Q$sigils{$reftype}\E/;

 $reftype;
}

sub _clean_msg {
 my ($msg) = @_;

 $msg =~ s/(?:\s+called)?\s+at\s+.*$//s;

 $msg;
}

=head2 C<flatten $proto, @args>

Flattens the array C<@args> according to the prototype C<$proto>.
When C<@args> is what C<@_> is after calling a subroutine with prototype C<$proto>, C<flatten> returns the list of what C<@_> would have been if there were no prototype.
It croaks if the arguments can't possibly match the required prototype, e.g. when a reference type is wrong or when not enough elements were provided.

=cut

sub flatten {
 my $proto = shift;

 return @_ unless defined $proto;

 my @args;
 while ($proto =~ /(\\?)(\[[^\]]+\]|[^\];])/g) {
  my $sigil = $2;

  if ($1) {
   my $arg     = shift;
   my $reftype = _check_ref $arg, $sigil;

   push @args, $reftype eq 'SCALAR'
               ? $$arg
               : ($reftype eq 'ARRAY'
                  ? @$arg
                  : ($reftype eq 'HASH'
                     ? %$arg
                     : ($reftype eq 'GLOB'
                        ? *$arg
                        : &$arg # _check_ref ensures this must be a code ref
                       )
                    )
                 );

  } elsif ($sigil =~ /[\@\%]/) {
   push @args, @_;
   last;
  } else {
   croak 'Not enough arguments to match this prototype' unless @_;
   push @args, shift;
  }
 }

 return @args;
}

=head2 C<wrap $name, %opts>

Generates a wrapper that calls the function C<$name> with a prototyped argument list.
That is, the wrapper's arguments should be what C<@_> is when you define a subroutine with the same prototype as C<$name>.

    my $a = [ 0 .. 2 ];
    my $push = wrap 'CORE::push';
    $push->($a, 3, 4); # returns 3 + 2 = 5 and $a now contains 0 .. 4

You can force the use of a specific prototype.
In this case, C<$name> must be a hash reference that holds exactly one key / value pair, the key being the function name and the value the prototpye that should be used to call it.

    my $push = wrap { 'CORE::push' => '\@$' }; # only pushes 1 arg

Others arguments are seen as key / value pairs that are meant to tune the code generated by L</wrap>.
Valid keys are :

=over 4

=item C<< ref => $func >>

Specifies the function used in the generated code to test the reference type of scalars.
Defaults to C<'ref'>.
You may also want to use L<Scalar::Util/reftype>.

=item C<< wrong_ref => $code >>

The code executed when a reference of incorrect type is encountered.
The result of this snippet is also the result of the generated code, hence it defaults to C<'undef'>.
It's a good place to C<croak> or C<die> too.

=item C<< sub => $bool >>

Encloses the code into a C<sub { }> block.
Default is true.

=item C<< compile => $bool >>

Makes L</wrap> compile the code generated and return the resulting code reference.
Be careful that in this case C<ref> must be a fully qualified function name.
Defaults to true, but turned off when C<sub> is false.

=back

For example, this allows you to recall into C<CORE::grep> and C<CORE::map> by using the C<\&@> prototype :

    my $grep = wrap { 'CORE::grep' => '\&@' };
    sub mygrep (&@) { $grep->(@_) } # the prototypes are intentionally different

=cut

sub _wrap {
 my ($name, $proto, $i, $args, $coderefs, $opts) = @_;

 while ($proto =~ s/(\\?)(\[[^\]]+\]|[^\];])//) {
  my ($ref, $sigil) = ($1, $2);
  $sigil = $1 if $sigil =~ /^\[([^\]]+)\]/;

  my $cur = "\$_[$i]";

  if ($ref) {
   if (length $sigil > 1) {
    my $code     = "my \$r = $opts->{ref}($cur); ";
    my @branches = map {
     my $subcall = _wrap(
      $name, $proto, ($i + 1), $args . "$_\{$cur}, ", $coderefs, $opts
     );
     "if (\$r eq '$reftypes{$_}') { $subcall }";
    } split //, $sigil;
    $code .= join ' els', @branches, "e { $opts->{wrong_ref} }";
    return $code;
   } else {
    $args .= "$sigil\{$cur}, ";
   }
  } elsif ($sigil =~ /[\@\%]/) {
   $args .= '@_[' . $i . '..$#_]';
  } elsif ($sigil =~ /\&/) {
   my %h = do { my $c; map { $_ => $c++ } @$coderefs };
   my $j;
   if (exists $h{$i}) {
    $j = int $h{$i};
   } else {
    push @$coderefs, $i;
    $j = $#{$coderefs};
   }
   $args .= "sub{&{\$c[$j]}}, ";
  } elsif ($sigil eq '_') {
   $args .= "((\@_ > $i) ? $cur : \$_), ";
  } else {
   $args .= "$cur, ";
  }
 } continue {
  ++$i;
 }

 $args =~ s/,\s*$//;

 return "$name($args)";
}

sub _check_name {
 my ($name) = @_;
 croak 'No subroutine specified' unless $name;

 my $proto;
 my $r = ref $name;
 if (!$r) {
  $proto = prototype $name;
 } elsif ($r eq 'HASH') {
  croak 'Forced prototype hash reference must contain exactly one key/value pair' unless keys %$name == 1;
  ($name, $proto) = %$name;
 } else {
  croak 'Unhandled ' . $r . ' reference as first argument';
 }

 $name =~ s/^\s+//;
 $name =~ s/[\s\$\@\%\*\&;].*//;

 return $name, $proto;
}

sub wrap {
 my ($name, $proto) = _check_name shift;
 croak 'Optional arguments must be passed as key => value pairs' if @_ % 2;
 my %opts = @_;

 $opts{ref}     ||= 'ref';
 $opts{sub}       = 1       unless defined $opts{sub};
 $opts{compile}   = 1       if     not defined $opts{compile} and $opts{sub};
 $opts{wrong_ref} = 'undef' unless defined $opts{wrong_ref};

 my @coderefs;
 my $call;
 if (defined $proto) {
  $call = _wrap $name, $proto, 0, '', \@coderefs, \%opts;
 } else {
  $call = _wrap $name, '', 0, '@_';
 }

 if (@coderefs) {
  my $decls = @coderefs > 1 ? 'my @c = @_[' . join(', ', @coderefs) . ']; '
                            : 'my @c = ($_[' . $coderefs[0] . ']); ';
  $call = $decls . $call;
 }

 $call = "{ $call }";
 $call = "sub $call" if $opts{sub};

 if ($opts{compile}) {
  my $err;
  {
   local $@;
   $call = eval $call;
   $err  = $@;
  }
  croak _clean_msg $err if $err;
 }

 return $call;
}

=head2 C<recall $name, @args>

Calls the function C<$name> with the prototyped argument list C<@args>.
That is, C<@args> should be what C<@_> is when you call a subroutine with C<$name> as prototype.
You can still force the prototype by passing C<< { $name => $proto } >> as the first argument.

    my $a = [ ];
    recall { 'CORE::push' => '\@$' }, $a, 1, 2, 3; # $a just contains 1

It's implemented in terms of L</wrap>, and hence calls C<eval> at each run.
If you plan to recall several times, consider using L</wrap> instead.

=cut

sub recall;

BEGIN {
 my $safe_wrap = sub {
  my $name = shift;

  my ($wrap, $err);
  {
   local $@;
   $wrap = eval { wrap $name };
   $err  = $@;
  }

  $wrap, $err;
 };

 if ("$]" == 5.008) {
  # goto tends to crash a lot on perl 5.8.0
  *recall = sub {
   my ($wrap, $err) = $safe_wrap->(shift);
   croak _clean_msg $err if $err;
   $wrap->(@_)
  }
 } else {
  *recall = sub {
   my ($wrap, $err) = $safe_wrap->(shift);
   croak _clean_msg $err if $err;
   goto $wrap;
  }
 }
}

=head1 EXPORT

The functions L</flatten>, L</wrap> and L</recall> are only exported on request, either by providing their name or by the C<':funcs'> and C<':all'> tags.

=cut

use base qw<Exporter>;

use vars qw<@EXPORT @EXPORT_OK %EXPORT_TAGS>;

@EXPORT             = ();
%EXPORT_TAGS        = (
 'funcs' =>  [ qw<flatten wrap recall> ]
);
@EXPORT_OK          = map { @$_ } values %EXPORT_TAGS;
$EXPORT_TAGS{'all'} = [ @EXPORT_OK ];

=head1 DEPENDENCIES

L<Carp>, L<Exporter> (core modules since perl 5), L<Scalar::Util> (since 5.7.3).

=head1 AUTHOR

Vincent Pit, C<< <perl at profvince.com> >>, L<http://www.profvince.com>.

You can contact me by mail or on C<irc.perl.org> (vincent).

=head1 BUGS

Please report any bugs or feature requests to C<bug-sub-prototype-util at rt.cpan.org>, or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Sub-Prototype-Util>.
I will be notified, and then you'll automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Sub::Prototype::Util

Tests code coverage report is available at L<http://www.profvince.com/perl/cover/Sub-Prototype-Util>.

=head1 COPYRIGHT & LICENSE

Copyright 2008,2009,2010,2011 Vincent Pit, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1; # End of Sub::Prototype::Util
